File: benchmarks/async/05_backpressure.py
"""Async benchmark 5: Backpressure chain (Python asyncio baseline)."""
import asyncio

async def stage_a(n: int) -> int:
    return n + 1

async def stage_b(n: int) -> int:
    return n * 2

async def stage_c(n: int) -> int:
    return n - 1

async def pipeline() -> int:
    total = 0
    for i in range(50):
        a = await stage_a(i)
        b = await stage_b(a)
        c = await stage_c(b)
        total += c
    return total

def main():
    result = asyncio.run(pipeline())
    print(f"checksum = {result}")

if __name__ == "__main__":
    main()

File: benchmarks/async/05_backpressure.go
package main

import "fmt"

func stageA(n int) <-chan int {
	ch := make(chan int, 1)
	go func() { ch <- n + 1 }()
	return ch
}

func stageB(n int) <-chan int {
	ch := make(chan int, 1)
	go func() { ch <- n * 2 }()
	return ch
}

func stageC(n int) <-chan int {
	ch := make(chan int, 1)
	go func() { ch <- n - 1 }()
	return ch
}

func main() {
	total := 0
	for i := 0; i < 50; i++ {
		a := <-stageA(i)
		b := <-stageB(a)
		c := <-stageC(b)
		total += c
	}
	fmt.Printf("checksum = %d\n", total)
}
